src/gt4py/next/iterator/type_inference.py

```diff
@@ -179,6 +179,83 @@ def handle_constraint(
         return False
 
 
+class ValListTuple(Type):
+    """
+    A tuple of `Val` that contains `List`s.
+
+    All items have
+    - the same `kind` and `size`;
+    - `dtype` is `List` with different `list_dtypes`, but same `max_length`, and `has_skip_values`.
+    """
+
+    kind: Type = eve.field(default_factory=TypeVar.fresh)
+    list_dtypes: Type = eve.field(default_factory=TypeVar.fresh)
+    max_length: Type = eve.field(default_factory=TypeVar.fresh)
+    has_skip_values: Type = eve.field(default_factory=TypeVar.fresh)
+    size: Type = eve.field(default_factory=TypeVar.fresh)
+
+    def __eq__(self, other: typing.Any) -> bool:
+        if (
+            isinstance(self.list_dtypes, Tuple)
+            and isinstance(self.max_length, Tuple)
+            and isinstance(self.has_skip_values, Tuple)
+            and isinstance(other, Tuple)
+        ):
+            list_dtypes: Type = self.list_dtypes
+            elems: Type = other
+            while (
+                isinstance(list_dtypes, Tuple)
+                and isinstance(elems, Tuple)
+                and Val(
+                    kind=self.kind,
+                    dtype=List(
+                        list_dtypes.front,
+                        max_length=self.max_length,
+                        has_skip_values=self.has_skip_values,
+                    ),
+                    size=self.size,
+                )
+                == elems.front
+            ):
+                list_dtypes = list_dtypes.others
+                elems = elems.others
+            return list_dtypes == elems == EmptyTuple()
+
+        return (
+            isinstance(other, ValListTuple)
+            and self.kind == other.kind
+            and self.list_dtypes == other.list_dtypes
+            and self.max_length == other.max_length
+            and self.has_skip_values == other.has_skip_values
+            and self.size == other.size
+        )
+
+    def handle_constraint(
+        self, other: Type, add_constraint: abc.Callable[[Type, Type], None]
+    ) -> bool:
+        if isinstance(other, Tuple):
+            list_dtypes = [TypeVar.fresh() for _ in other]
+            max_length = TypeVar.fresh()
+            has_skip_values = TypeVar.fresh()
+            expanded = [
+                Val(
+                    kind=self.kind,
+                    dtype=List(dtype=dtype, max_length=max_length, has_skip_values=has_skip_values),
```

Comment:
You can directly use `self.max_length` and `self.has_skip_values` here.
```suggestion
                    dtype=List(dtype=dtype, max_length=self.max_length, has_skip_values=self.has_skip_values),
```

Reply:
done